modules/dex-oracle/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+// This file is part of Acala.
+
+// Copyright (C) 2020-2022 Acala Foundation.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! # DEX Oracle Module
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+
+use frame_support::{pallet_prelude::*, traits::Time, transactional};
+use frame_system::pallet_prelude::*;
+use orml_traits::Happened;
+use primitives::{Balance, CurrencyId, TradingPair};
+use sp_core::U256;
+use sp_runtime::{traits::Saturating, FixedPointNumber, SaturatedConversion};
+use sp_std::marker::PhantomData;
+use support::{DEXManager, DEXPriceProvider, ExchangeRate};
+
+mod mock;
+mod tests;
+//pub mod weights;
+
+pub use module::*;
+//pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	pub type MomentOf<T> = <<T as Config>::Time as Time>::Moment;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// DEX provide liquidity info.
+		type DEX: DEXManager<Self::AccountId, CurrencyId, Balance>;
+
+		/// Time provider
+		type Time: Time;
+
+		/// The origin which may manage dex oracle.
+		type UpdateOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The time interval in millisecond for updating the cumulative prices.
+		#[pallet::constant]
+		type IntervalToUpdateCumulativePrice: Get<MomentOf<Self>>;
```

Comment:
I am not sure if we want to use time or just block number. Which one is safer when block stopped for 30 mins?

Reply:
Maybe uniswap has considered this, TWAP stands for the time-weighted average price, the price is still available for the duration of that unstable block.  I prefer use time as the weight